### ISSUE TITLE
Strange gradients.

### ISSUE THREAD

**Issue by airalcorn2** [daniilidis-group/neural_renderer#18]
Hi. My colleague (@czw0078) and I have been using your neural renderer port, and we've noticed some pretty strange behavior during optimization (see also, [this issue](https://github.com/hiroharu-kato/neural_renderer/issues/9)). A minimal working example demonstrating the odd behavior can be found below. The code requires adding:
```python
if up is None:
    up = torch.cuda.FloatTensor([0, 1, 0])
```
to `look.py` to run.
```python
import neural_renderer as nr
import numpy as np
import scipy.misc
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self, input_obj, initial_z):
        super(Model, self).__init__()

        # Load object.
        (vertices, faces) = nr.load_obj(input_obj)

        self.vertices = vertices[None, :, :]
        self.faces = faces[None, :, :]
        texture_size = 2
        self.textures = torch.ones(1, self.faces.shape[1], texture_size, texture_size,
                                   texture_size, 3, dtype=torch.float32).cuda()

        self.x = self.vertices[0][:, 0]
        self.y = self.vertices[0][:, 1]
        self.z = self.vertices[0][:, 2]

        # Camera parameters.
        camera_distance = 2.732
        elevation = 0
        azimuth = 0
        camera_position = np.array(nr.get_points_from_angles(camera_distance, elevation, azimuth), dtype=np.float32)
        self.camera_position = torch.from_numpy(camera_position).cuda()

        # Adjust renderer.
        renderer = nr.Renderer(camera_mode="look")
        renderer.eye = self.camera_position
        renderer.viewing_angle = 8.123
        self.renderer = renderer

        # Optimization direction.
        self.z_delta = nn.Parameter(torch.Tensor([initial_z]))

    def forward(self):
        new_z = self.z + self.z_delta

        vertices = torch.stack((self.x, self.y, new_z), dim=1)[None, :, :]
        image = self.renderer(vertices, self.faces, self.textures)

        return (new_z, image)


def main():
    input_obj = "./data/teapot.obj"
    model = Model(input_obj, 4)
    model.cuda()

    gen_teapot = False

    for sign in [1, -1]:
        model.zero_grad()
        (new_z, images) = model()

        if gen_teapot and sign == 1:
            image = images.detach().cpu().numpy()[0].transpose(1, 2, 0)
            scipy.misc.toimage(image, cmin=0, cmax=1).save("my_teapot.png")

        print("Current z_delta: {0}".format(model.z_delta.item()))
        assert (model.z_delta.grad is None) or (model.z_delta.grad.item() == 0.0)

        loss = sign * images.norm()
        loss.backward()
        print("Loss: {0}".format(loss.item()))

        print("z_delta derivative after .backward(): {0}".format(model.z_delta.grad[0]))


if __name__ == "__main__":
    main()
```
The example code produces the following output:
```
Current z_delta: 4.0
Loss: 146.864685059
z_delta derivative after .backward(): -232.424438477
Current z_delta: 4.0
Loss: -146.864685059
z_delta derivative after .backward(): -46.2437057495
```
when using the image below as a starting point.
![my_teapot](https://user-images.githubusercontent.com/1927770/44945469-d63cf000-adaf-11e8-84aa-9bce93c54d62.png)
As you can see, the gradients have the same sign despite using opposite loss functions. Any insight you could provide on this behavior would be greatly appreciated. Thank you.

The following `Dockerfile` was used to generate the environment used for the code above.
```docker
FROM nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04
RUN apt-get update && apt-get install -y --no-install-recommends \
         wget \
         build-essential \
         cmake \
         nano \
         less \
         git \
         curl \
         libjpeg-dev \
         libpng-dev \
         imagemagick \
         python \
         python-dev \
         python-setuptools \
         python-pip \
         python-wheel
RUN pip install torch torchvision scikit-image tqdm imageio ipython==5.8.0
```

**Comment by nkolot** [comment]
Ooops! That is indeed very weird. So it is an issue of the original Chainer implementation too. I will have to look into the equations of the paper and the CUDA code to try and figure out what is happening here.

We should expect that the gradients before and after should have the same magnitude and flipped sign.

**Comment by nkolot** [comment]
I've located the source of assymetry. If you look at the lines (401,402) and (481,482) in [this](https://github.com/daniilidis-group/neural_renderer/blob/master/neural_renderer/cuda/rasterize_cuda_kernel.cu) file, there is a statement
```
if (diff_grad <= 0)
 continue;
```

If you comment this out, then the 2 gradient values have the same magnitude and opposite signs. However I feel that this "hack" is not correct, even though it still passes all the test cases. Probably the code in this part needs to be written again from scratch. It is not documented well by the authors so it won't be trivial to figure out how it works.

Nevertheless, can you try to run it with this fix and see if your results improve?

**Comment by airalcorn2** [comment]
Thanks for looking into it, @nkolot. Initial results seem encouraging.
**Positive Loss/Blacken**
![teapot_z_blacken](https://user-images.githubusercontent.com/1927770/44993054-cc0d2400-af5e-11e8-8ebf-4acb3188f773.gif)
**Negative Loss/Whiten**
![teapot_z_whiten](https://user-images.githubusercontent.com/1927770/44993055-cc0d2400-af5e-11e8-9cdb-0e4ce66f6e85.gif)
Agreed that the fix feels "hack-y", but it's hard to understand why the conditional is there in the first place because of the lack of documentation. My best guess is it's related to equations **_(1)_**, **_(3)_**, and **_(4)_** in the paper, where they set the partial derivative to zero when `delta_P_j * delta_I_j >= zero`, but the conditional in the code doesn't reference `delta_P_j` and is `<=` rather than `>=`. Very strange!

**Comment by nkolot** [comment]
I am reopening the issue, just to keep track that I have to properly fix this and submit a PR. Will do it when I find time, probably later in the week.
